kernel/src/client/default/parquet.rs

```diff
@@ -120,26 +121,41 @@ impl FileOpener for ParquetOpener {
 
         let batch_size = self.batch_size;
         // let projection = self.projection.clone();
-        let _table_schema = self.table_schema.clone();
+        let table_schema = self.table_schema.clone();
         let limit = self.limit;
 
         Ok(Box::pin(async move {
             // TODO avoid IO by converting passed file meta to ObjectMeta
             let meta = store.head(&path).await?;
-            let reader = ParquetObjectReader::new(store, meta);
+            let mut reader = ParquetObjectReader::new(store, meta);
+            let metadata = ArrowReaderMetadata::load_async(&mut reader, Default::default()).await?;
+            let parquet_schema = metadata.schema();
+            let indicies = get_requested_indicies(&table_schema, parquet_schema)?;
             let options = ArrowReaderOptions::new(); //.with_page_index(enable_page_index);
             let mut builder =
                 ParquetRecordBatchStreamBuilder::new_with_options(reader, options).await?;
+            if let Some(mask) = generate_mask(
+                &table_schema,
+                parquet_schema,
+                builder.parquet_schema(),
+                &indicies,
+            ) {
+                builder = builder.with_projection(mask)
+            }
 
-            // let mask = ProjectionMask::roots(builder.parquet_schema(), projection.iter().cloned());
             if let Some(limit) = limit {
                 builder = builder.with_limit(limit)
             }
 
-            let stream = builder
-                // .with_projection(mask)
-                .with_batch_size(batch_size)
-                .build()?;
+            let stream = builder.with_batch_size(batch_size).build()?;
+
+            let stream = stream.map(move |rbr| {
+                // re-order each batch if needed
+                rbr.and_then(|rb| {
+                    reorder_record_batch(table_schema.clone(), rb, &indicies)
+                        .map_err(|e| parquet::errors::ParquetError::General(format!("{e}")))
```

Comment:
the error handling is odd here. i'm working on getting it less convoluted and making the compiler happy.